### PR TITLE
Factor out `MultiWriter` token from `RoomStreamToken`

### PR DESCRIPTION
**Pull request by erikjohnston**
So we can reuse code when we split up other streams.

### CODE REVIEW

File: synapse/types/__init__.py [hunk 1]
```diff
@@ -60,6 +61,8 @@
 from synapse.util.stringutils import parse_and_validate_server_name
 
 if TYPE_CHECKING:
+    from typing_extensions import Self
```
Comment (erikjohnston) [inline synapse/types/__init__.py:64]: This requires `typing_extensions >= 4.0`. Not sure how we feel about bumping that? Or just leaving it behind `TYPE_CHECKING`?

Comment (DMRobertson) [reply synapse/types/__init__.py:64]: I'm happy as long as the packagers are. https://pkgs.org/search/?q=typing-extensions and https://repology.org/project/python:typing-extensions/versions. Debian buster and Ubuntu focal + jammy have 3.x; I can't remember if we've dropped support for those.

Comment (DMRobertson) [reply synapse/types/__init__.py:64]: Oh ISWYM. It'd be nice to be able to just import it, but I don't mind the conditional import and referencing it in "quotes" if that helps the packagers too. (ISTR we have to quote "defer.Deferred[blah]" anyway...

Comment (clokep) [reply synapse/types/__init__.py:64]: I think putting it behind `TYPE_CHECKING` is fine.

File: synapse/types/__init__.py [hunk 2]
```diff
@@ -592,7 +658,9 @@ def copy_and_advance(self, other: "RoomStreamToken") -> "RoomStreamToken":
             for instance in set(self.instance_map).union(other.instance_map)
         }
 
-        return RoomStreamToken(None, max_stream, immutabledict(instance_map))
+        return RoomStreamToken(
+            stream=max_stream, instance_map=immutabledict(instance_map)
+        )
```
Comment (clokep) [inline synapse/types/__init__.py:662]: I think most of this is `super().copy_and_advance(...)`?

File: synapse/types/__init__.py [hunk 3]
```diff
@@ -436,7 +439,78 @@ def f2(m: Match[bytes]) -> bytes:
...
+    stream: int = attr.ib(validator=attr.validators.instance_of(int), kw_only=True)
+
+    instance_map: "immutabledict[str, int]" = attr.ib(
+        factory=immutabledict,
+        validator=attr.validators.deep_mapping(
+            key_validator=attr.validators.instance_of(str),
+            value_validator=attr.validators.instance_of(int),
+            mapping_validator=attr.validators.instance_of(immutabledict),
+        ),
+        kw_only=True,
```
Comment (clokep) [inline synapse/types/__init__.py:460]: OMG can we use `kw_only` everywhere?!?! 😍

**Review by clokep — APPROVED** [review]
Seems legit if CI is green.

File: synapse/types/__init__.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/types/__init__.py]: whoops, that was what I was planning on doing

**Pull request merged**